[Feature] Add Kibana security roles

nimdeveloper opened the pull request:
Closes #406

nimdeveloper commented:
> <!-- CLA-CHECK:435 -->
> &#10060; Author of the following commits did not sign a [Contributor Agreement](https://www.elastic.co/contributor-agreement):
>    5e4ed503ed9bca2f8fb458cb221859692b569abd, 26141bdad5bb651d877167528455a7fcefde8483
> 
>    Please, read and sign the above mentioned agreement if you want to contribute to this project

Signed.

tobio commented on docs/data-sources/kibana_security_role.md:
```diff
@@ -0,0 +1,96 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "elasticstack_kibana_security_role Data Source - terraform-provider-elasticstack"
+subcategory: ""
```
Can we create docs [templates](https://github.com/elastic/terraform-provider-elasticstack/blob/main/templates/resources/kibana_alerting_rule.md.tmpl) for these new resources with the subcategory set to Kibana.

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+						MaxItems:    1,
+						Elem: &schema.Schema{
+							Type: schema.TypeString,
+							ValidateDiagFunc: func(v any, p cty.Path) diag.Diagnostics {
+								value := v.(string)
+								expected := []string{"all", "read"}
+								var diags diag.Diagnostics
+								for _, e := range expected {
+									if e == value {
+										return diags
+									}
+								}
+								diag := diag.Diagnostic{
+									Severity: diag.Error,
+									Summary:  "Wrong value for base attribute",
+									Detail:   fmt.Sprintf("Expected %s , got %s", strings.Join(expected, " | "), value),
+								}
+								diags = append(diags, diag)
+								return diags
+							},
```
```suggestion
							ValidateFunc: validation.StringInSlice([]string{"all", "read"}),
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+						Description: "A base privilege. When specified, the base must be [\"all\"] or [\"read\"].",
+						Type:        schema.TypeSet,
+						Optional:    true,
+						MaxItems:    1,
```
We should be able to do something like 

```suggestion
						MaxItems:    1,
						ExactlyOneOf: []string{"kibana.base", "kibana.feature"},
```

nimdeveloper replied:
This is not possible due to plugin restrictions,  
[https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema#Schema](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema#Schema)  
As we use TypeSet for `"kibana"` and `"kibana.feature"` , based on the readme above, it's not possible.

  
Related:  
hashicorp/terraform-plugin-sdk#1001

hashicorp/terraform-plugin-sdk#71

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+					"feature": {
+						Description: "List of privileges for specific features. When the feature privileges are specified, you are unable to use the \"base\" section.",
+						Type:        schema.TypeSet,
+						Optional:    true,
```
Similarly

```suggestion
						Optional:    true,
						ExactlyOneOf: []string{"kibana.base", "kibana.feature"},
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+			}
+
+			if basePrivileges, ok := each["base"].(*schema.Set); ok && basePrivileges.Len() > 0 {
+				if _features, ok := each["feature"].(*schema.Set); ok && _features.Len() > 0 {
+					return diag.Errorf("Only one of the `feature` or `base` privileges allowed!")
+				}
```
The `ExactlyOneOf` above should make this unnecessary.

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+		kibanaRole.Kibana = make([]kbapi.KibanaRoleKibana, definedKibanaConfigs.Len())
+		for i, item := range definedKibanaConfigs.List() {
+			each := item.(map[string]interface{})
+			_config := kbapi.KibanaRoleKibana{
```
```suggestion
			config := kbapi.KibanaRoleKibana{ 
```

Can we remove the leading underscores

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+					}
+					_config.Feature[featureData["name"].(string)] = _features
+				}
+			} else {
+				return diag.Errorf("Either on of the `feature` or `base` privileges must be set for kibana role!")
+			}
```
Similarly this should be covered by the `ExactlyOneOf` validation.

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+					return diag.Errorf("Only one of the `feature` or `base` privileges allowed!")
+				}
+				_config.Base = make([]string, basePrivileges.Len())
+				for i, name := range basePrivileges.List() {
+					_config.Base[i] = name.(string)
+				}
```
We'd prefer to just append to a slice rather than assign an index.

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+
+	if v, ok := d.GetOk("elasticsearch"); ok {
+		if definedElasicConfigs := v.(*schema.Set); definedElasicConfigs.Len() > 0 {
+			userElasitcConfig := definedElasicConfigs.List()[0].(map[string]interface{})
```
```suggestion
			userElasticConfig := definedElasticConfigs.List()[0].(map[string]interface{})
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+		return diag.FromErr(err)
+	}
+
+	id, diags := client.ID(ctx, roleManageResponse.Name)
+	if diags.HasError() {
+		return diags
+	}
+
+	d.SetId(id.String())
```
I know you've just copied it, but this creates a dependency on Elasticsearch being configured when managing a Kibana resource. Let's just use the role name as the ID
```suggestion
	d.SetId(roleManageResponse.Name)
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+	if diags.HasError() {
+		return diags
+	}
+	compId, diags := clients.CompositeIdFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+	name := compId.ResourceId
```
```suggestion
	name := d.ID()
```

tobio commented on examples/resources/elasticstack_kibana_security_role/import.sh:
```diff
@@ -0,0 +1 @@
+terraform import elasticstack_kibana_security_role.example_role <cluster_uuid>/<role name>
```
```suggestion
terraform import elasticstack_kibana_security_role.example_role <role name>
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,506 @@
+	if v, ok := d.GetOk("elasticsearch"); ok {
+		if definedElasicConfigs := v.(*schema.Set); definedElasicConfigs.Len() > 0 {
+			userElasitcConfig := definedElasicConfigs.List()[0].(map[string]interface{})
+			if v, ok := userElasitcConfig["cluster"]; ok {
```
There's a lot going on in here. It would be nicer to break some of these out to `expand<Thing>` functions.

tobio commented on internal/kibana/role_data_source.go:
```diff
@@ -0,0 +1,180 @@
+		},
+	}
+
+	utils.AddConnectionSchema(roleSchema)
```
```suggestion
```

This is ES specific, and shouldn't be added to new resources.

tobio reviewed (commented):
Thanks for this, added some comments inline. 

Can we also add an entry to the [changelog](https://github.com/elastic/terraform-provider-elasticstack/blob/main/CHANGELOG.md) for this PR.

nimdeveloper commented:
> Thanks for this, added some comments inline.
> 
> Can we also add an entry to the [changelog](https://github.com/elastic/terraform-provider-elasticstack/blob/main/CHANGELOG.md) for this PR.

Thanks for the reply, I updated the reviews mentioned.

tobio commented on CHANGELOG.md:
```diff
@@ -1,4 +1,5 @@
 ## [Unreleased]
+- Add support for Kibana security role ([#435](https://github.com/elastic/terraform-provider-elasticstack/pull/435))
 
 ### Added
 - Introduce `elasticstack_kibana_import_saved_objects` resource as an additive only way to manage Kibana saved objects ([#343](https://github.com/elastic/terraform-provider-elasticstack/pull/343)).
```
```suggestion

### Added
- Add support for Kibana security role ([#435](https://github.com/elastic/terraform-provider-elasticstack/pull/435))
- Introduce `elasticstack_kibana_import_saved_objects` resource as an additive only way to manage Kibana saved objects ([#343](https://github.com/elastic/terraform-provider-elasticstack/pull/343)).
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+)
+
+func ResourceRole() *schema.Resource {
+	apikeySchema := map[string]*schema.Schema{
```
```suggestion
	roleSchema := map[string]*schema.Schema{
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: apikeySchema,
```
```suggestion
		Schema: roleSchema,
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+
+// Helper functions
+
+func expandMetadata(v interface{}) (map[string]interface{}, diag.Diagnostics) {
```
These will be visible through the entire `internal/kibana` package. It'd be worthwhile to have some more specific naming here. 

```suggestion
func expandKibanaRoleMetadata(v interface{}) (map[string]interface{}, diag.Diagnostics) {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return metadata, nil
+}
+
+func expandElasticsearch(v interface{}) *kbapi.KibanaRoleElasticsearch {
```
```suggestion
func expandKibanaRoleElasticsearch(v interface{}) *kbapi.KibanaRoleElasticsearch {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return elasticConfig
+}
+
+func expandKibana(v interface{}) ([]kbapi.KibanaRoleKibana, diag.Diagnostics) {
```
```suggestion
func expandKibanaRoleKibana(v interface{}) ([]kbapi.KibanaRoleKibana, diag.Diagnostics) {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return kibanaConfigs, nil
+}
+
+func flattenIndicesData(indices *[]kbapi.KibanaRoleElasticsearchIndice) []interface{} {
```
```suggestion
func flattenKibanaRoleIndicesData(indices *[]kbapi.KibanaRoleElasticsearchIndice) []interface{} {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return make([]interface{}, 0)
+}
+
+func flattenElasticsearchData(elastic *kbapi.KibanaRoleElasticsearch) []interface{} {
```
```suggestion
func flattenKibanaRoleElasticsearchData(elastic *kbapi.KibanaRoleElasticsearch) []interface{} {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return make([]interface{}, 0)
+}
+
+func flattenKibanaFeatureData(features map[string][]string) []interface{} {
```
```suggestion
func flattenKibanaRoleKibanaFeatureData(features map[string][]string) []interface{} {
```

tobio commented on internal/kibana/role.go:
```diff
@@ -0,0 +1,501 @@
+	return make([]interface{}, 0)
+}
+
+func flattenKibanaData(kibana_configs *[]kbapi.KibanaRoleKibana) []interface{} {
```
```suggestion
func flattenKibanaRoleKibanaData(kibana_configs *[]kbapi.KibanaRoleKibana) []interface{} {
```

tobio reviewed (commented):
LGTM. Some minor naming nits.

tobio reviewed (approved):
Lgtm, this is awesome thanks!
